rascal-textmate-core/src/main/rascal/lang/rascal/grammar/Util.rsc

```diff
@@ -47,6 +49,62 @@ bool isRecursive(Grammar g, Symbol s) {
     return check({}, s);
 }
 
+@synopsis{
+    Representation of a pointer to a symbol in (the list of symbols of) a
+    production. This is useful to distinguish between different occurrences of
+    the same symbol in a grammar (i.e., they have different pointers).
+}
+
+alias Pointer = tuple[Production p, int index];
+
+@synopsis{
+    Finds the list of pointers -- a *trace* -- to the first occurrence of symbol
+    `s`, if any, starting from production `p`, optionally in a particular
+    direction (default: `forward()`). That is: if `<p1,i>` is followed by
+    `<p2,_>` in the returned list, then `p1.symbols[i]` is a non-terminal and
+    `p2` is one of its productions.
+}
+
+@description{
+    For instance, consider the following grammar:
+
+    ```
+    lexical X  = Y;
+    lexical Y  = alt1: "[" "[" "[" Z1 "]" "]" "]" | alt2: "<" Z2 ">"; 
+    lexical Z1 = "foo" "bar";
+    lexical Z2 = "baz";
+    ```
+
+    The list of pointers to `"bar"`, starting from `X`, is:
+
+      - `<X,0>`
+      - `<Y.alt1,3>`
+      - `<Z1,1>`
+    
+    The list of pointers to `"qux"` is just empty.
+}
+
+list[Pointer] find(Grammar g, Production p, Symbol s, Direction dir = forward()) {
+
+    list[Pointer] doFind(set[Production] doing, Production haystack, Symbol needle) {
+        for (haystack notin doing, i <- reorder([0..size(haystack.symbols)], dir)) {
+            Symbol ith = delabel(haystack.symbols[i]);
+            if (ith == needle) {
+                return [<haystack, i>];
+            }
+            for (isNonTerminalType(ith), child <- lookup(g, ith)) {
+                if (list[Pointer] l: [_, *_] := doFind(doing + haystack, child, s)) {
+                    return [<haystack, i>] + l;
+                }
+            }
+        }
+
+        return [];
+    }
+
+    return doFind({}, p, s);
```

Comment:
is there a possibility of making a graph out of the grammar and using the graph features of rascal? (or `analysis::graphs` modules?)

Reply:
Definitely a possibility, probably 😉 I'll need to think a bit more carefully about both the details and the bigger picture, though, to get it right/better in other places too. For me, currently, the "smell" is that as the converter gets more complicated, I'm often calling `lookup` to expand non-terminals, so I'm starting to believe I should do some precomputation upfront to get the grammar in a more convenient representation (which could be a graph of some sort) that supports all these use cases. I'll address this in a separate PR.